### DISCUSSION TITLE
Switch to second window

### DISCUSSION THREAD

**Discussion by AliasTest** [cypress-io/cypress#29879] [Questions and Help]
I have below test scenario to automate 

Login to application and click on a task button that will open a task screen in another window with same session of 1st window.
Can I do that scenario using cypress?

**Comment by portyu9** [comment]
Hi @AliasTest 

Yes. For current Cypress, if you genuinely need to interact with a second browser window/tab, use the official `@cypress/puppeteer` plugin.

Cypress itself still runs in a single primary tab, but the Puppeteer plugin can connect to the Cypress-launched browser and interact with the additional window as another Puppeteer `Page`.

Set it up in `cypress.config.ts`:

```ts
import { defineConfig } from 'cypress'
import { setup, retry } from '@cypress/puppeteer'

export default defineConfig({
  e2e: {
    setupNodeEvents(on) {
      setup({
        on,
        onMessage: {
          async useTaskWindow(browser) {
            const page = await retry(async () => {
              const pages = await browser.pages()

              const taskPage = pages.find((page) =>
                page.url().includes('/task')
              )

              if (!taskPage) {
                throw new Error('Task window has not opened yet')
              }

              return taskPage
            })

            await page.bringToFront()

            // Interact/assert with the second window here using Puppeteer.
            return page.url()
          },
        },
      })
    },
  },
})
```

In the Cypress support file:

```ts
import '@cypress/puppeteer/support'
```

Then:

```js
cy.get('[data-cy=task]').click()

cy.puppeteer('useTaskWindow')
  .should('include', '/task')
```

Because it is the same browser instance, the second window can use the browser session/cookies established by the first window, subject to the application's normal cookie/origin rules.

If the fact that it opens a separate window is **not** itself important to the behavior being tested, a simpler Cypress test is often better: verify the target URL and visit it in the Cypress tab instead.

Also note that `@cypress/puppeteer` is currently beta and supports Chromium-based browsers. For headed Chrome-branded versions 137+, the plugin documentation recommends Chrome for Testing, Chromium, or Electron because of Chrome's extension-loading change.

No screenshots or attachments were included in this discussion.

Please mark this answer as accepted if it helped, thank you.
